controllers/global_search_setup.go

```diff
@@ -0,0 +1,553 @@
+	}
+
+	// 2. Enable federated search feature in the search-api deployment.
+	err = r.updateSearchApiDeployment(ctx, true, instance)
+	if err != nil {
+		log.Error(err, "Failed to enable the federated global search feature.")
+		return err // QUESTION: Should we return here or continue? Should we rollback other changes?
+	}
+
+	// Create configuration resources for each Managed Hub.
+	// MANAGED_HUBS=($(oc get managedcluster -o json | jq -r '.items[] | select(.status.clusterClaims[] |
+	//     .name == "hub.open-cluster-management.io" and .value != "NotInstalled") | .metadata.name'))
+	clusterList, err := r.DynamicClient.Resource(managedClusterResourceGvr).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		log.Error(err, "Failed to list of ManagedClusters to configure global search.")
+		return err // QUESTION: Should we return here or continue? Should we rollback other changes?
+	}
+
+	for _, cluster := range clusterList.Items {
+		isManagedHub := false
+		if cluster.Object["status"] == nil || cluster.Object["status"].(map[string]interface{})["clusterClaims"] == nil {
+			log.V(5).Info("Cluster doesn't have status or clusterClaims.", "cluster", cluster.GetName())
+			continue
+		}
+		clusterClaims := cluster.Object["status"].(map[string]interface{})["clusterClaims"].([]interface{})
+		for _, claim := range clusterClaims {
+			claimMap := claim.(map[string]interface{})
+			if claimMap["name"] == "hub.open-cluster-management.io" && claimMap["value"] != "NotInstalled" {
+				isManagedHub = true
+				break
+			}
+		}
+		if !isManagedHub {
+			log.V(5).Info("Cluster is not a Managed Hub.", "name", cluster.GetName())
+			continue
+		}
+
+		log.V(2).Info("Cluster is a Managed Hub. Configuring global search resources.", "name", cluster.GetName())
+
+		// 3a. Create a ManagedServiceAccount search-global.
+		_, err := r.DynamicClient.Resource(managedServiceAccountGvr).Namespace(cluster.GetName()).
+			Get(ctx, searchGlobal, metav1.GetOptions{})
+		if err != nil && errors.IsNotFound(err) {
+			log.V(1).Info("Creating ManagedServiceAccount search-global for managed hub", "name", cluster.GetName())
+			managedSA := &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"apiVersion": "authentication.open-cluster-management.io/v1beta1",
+					"kind":       "ManagedServiceAccount",
+					"metadata": map[string]interface{}{
+						"name": searchGlobal,
+						"labels": map[string]interface{}{
+							"app": "search",
+						},
+					},
+					"spec": map[string]interface{}{
+						"rotation": map[string]interface{}{},
+					},
+				},
+			}
+			_, err := r.DynamicClient.Resource(managedServiceAccountGvr).Namespace(cluster.GetName()).
+				Create(ctx, managedSA, metav1.CreateOptions{})
+			if err != nil {
+				log.Error(err, "Failed to create ManagedServiceAccount search-global.")
+				// QUESTION: How should we handle partial errors? Should we continue or rollback?
+			}
+		} else {
```

Comment:
There might be other error conditions where the client was not able to `Get` the resource - where  !errors.IsNotFound(err).

Reply:
Here I'm trying to create it only if not found, but maybe it's better if I try to create it and handle the case where it already exists.

Reply:
I reverted the logic. Now I try to create and ignore the error if already exists.
This is simpler and always 1 API request.

Reply:
That works great!